src/theory/builtin/theory_builtin_rewriter.cpp

```diff
@@ -70,6 +77,183 @@ Node TheoryBuiltinRewriter::blastChain(TNode in) {
   }
 }
 
+
+// recursion depth is bounded by the number of arguments in the bound variable list bvl
+Node TheoryBuiltinRewriter::getLambdaForArrayRepresentationRec( TNode a, TNode bvl, unsigned bvlIndex, 
+                                                                std::unordered_map< TNode, Node, TNodeHashFunction >& visited ){
+  std::unordered_map< TNode, Node, TNodeHashFunction >::iterator it = visited.find( a );
```

Comment:
The code in these functions is technical enough that a high-level comment explaining how to convert from LAMBDAs to arrays and back would be helpful.  A paragraph or two with a couple of examples should do it.  Be sure to include an example with more than one bound variable to show how that works.  This can go in the header file.